pkg/storagev2/migration/index.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package migration
+
+import (
+	storage "github.com/ethersphere/bee/pkg/storagev2"
+)
+
+type (
+	// ItemDeleteFn is callback function called in migration step
+	// to check if Item should be removed in this step.
+	ItemDeleteFn func(storage.Item) (delete bool)
+
+	// ItemUpdateFn is callback function called in migration step
+	// to check if Item should be updated in this step.
+	ItemUpdateFn func(storage.Item) (updateditem storage.Item, hasChanged bool)
+)
+
+// WithItemDeleteFn return option with ItemDeleteFn set.
+func WithItemDeleteFn(fn ItemDeleteFn) option {
+	return func(o *options) {
+		o.deleteFn = fn
+	}
+}
+
+// WithItemUpdaterFn return option with ItemUpdateFn set.
+func WithItemUpdaterFn(fn ItemUpdateFn) option {
+	return func(o *options) {
+		o.updateFn = fn
+	}
+}
+
+type option func(*options)
+
+type options struct {
+	deleteFn ItemDeleteFn
+	updateFn ItemUpdateFn
+}
+
+func newOptions() *options {
+	return &options{
+		deleteFn: func(storage.Item) bool { return false },
+		updateFn: func(i storage.Item) (storage.Item, bool) { return i, false },
+	}
+}
+
+func (o *options) applyAll(opts []option) {
+	for _, opt := range opts {
+		opt(o)
+	}
+}
+
+// NewStepOnIndex creates new migration step with update and/or delete operation.
+// Migration will iterate on all elements selected by query and delete or update items
+// based on supplied callback functions.
+func NewStepOnIndex(query storage.Query, opts ...option) StepFn {
+	o := newOptions()
+	o.applyAll(opts)
+
+	return func(s storage.Store) error {
+		itemsForDelete := make([]storage.Item, 0)
```

Comment:
You can avoid the zero length by declaring nil slice instead, which is the [preferred method](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices).